[SOT] Add fully `SizeVariable` support

LittleHeroZZZX opened the pull request:
### PR Category

User Experience

### PR Types

Improvements

### Description

### 1. 完善 SOT 中的 SizeVariable 支持

在 PR #76361 中，`paddle.shape` 的返回类型从 `list` 修改为继承自 `list` 的 `paddle.Size`。修改后，二者在行为上存在若干差异。为了确保 SOT (Symbolic Opcode Translator) 能够正确支持这些行为差异并避免类型降级或 BreakGraph，本 PR 对 `SizeVariable` 进行了完整的实现与适配。

`paddle.Size` 相比 `list` 的主要差异包括：

* **切片 (Slicing)**：返回 `paddle.Size` 类型而非 `list`。
* **拼接 (Concat)**：返回 `paddle.Size` 类型，且支持与 `tuple` 进行拼接。
* **重复 (Repeat)**：返回 `paddle.Size` 类型。
* **比较 (Compare)**：支持与 `tuple` 进行相等性比较（`paddle.Size([1, 2]) == (1, 2)` 为 `True`）。
* **Numel 方法**：新增 `numel()` 方法，返回所有元素的乘积。

本 PR 的具体修改点包括：

1. **新增 `SizeVariable` 类**：
* 继承自 `ListVariable`，重写了 `getitem`, `concat`, `repeat` 等方法，确保返回类型动态绑定为 `self.__class__`，防止类型降级为 `ListVariable`。
* 实现了 `numel` 方法的 Graph 逻辑。
* 实现了 `_reconstruct` 方法，确保还原为 Python 对象时生成 `paddle.Size` 指令。



3. **完善 Operator 注册**：
* 注册了 `paddle.Size` 的构造函数（支持 `ContainerVariable` 和 `IterVariable`）。
* 注册了 `operator.add`：支持 `Size + List/Tuple/Size` 以及 `List + Size`，结果均保持为 `SizeVariable`。
* 注册了 `operator.mul`：支持 `Size * int` 和反向乘法 `int * Size`。
* 注册了 `operator.eq/ne`：实现了 `SizeVariable` 与 `TupleVariable` 的内容比较逻辑，对齐 Paddle 动态图行为。


4. **添加单元测试**：
* 新增 `test_25_size.py`，覆盖了构造、切片、四则运算、反向运算、比较操作及 `numel` 方法，通过 `assert_results` 严格验证数值与类型正确性。



### 2. 优化 ListVariable

* 修改 `ListVariable` 的 `get_py_value`，使其返回 `self.get_py_type()` 而非硬编码的 `list`，从而支持子类（如 `SizeVariable`）正确返回其 Python 类型。
* 优化 `concat`, `repeat`, `getitem` 等方法，使用 `self.__class__` 替代硬编码的构造函数，增强代码复用性与多态支持。

LittleHeroZZZX committed:
feat(sot): enhance dispatcher with type scoring and add SizeVariable support

- Refactor dispatcher to handle multiple matching patterns by calculating type distance scores
- Add SizeVariable class and register paddle.Size constructor for iterables and containers
- Implement operator support for SizeVariable (mul, add, eq, ne) with list/tuple compatibility
- Fix reverse multiplication order for ConstantVariable * ListVariable/TupleVariable

SigureMo commented on python/paddle/jit/sot/opcode_translator/executor/variables/callable.py:
```diff
@@ -1316,6 +1316,31 @@ def from_value(value: Any, graph: FunctionGraph, tracker: Tracker):
+        from ..function_graph import convert_to_py_value
+
+        py_args = convert_to_py_value(args)
+        py_kwargs = convert_to_py_value(kwargs)
```
要转成 value 么？shape 里出现 symbol 是很常见的，这会导致 symbol 值被 guard 住

LittleHeroZZZX replied:
Done, 通过 get_wrapped_items 来构造

SigureMo commented on python/paddle/jit/sot/opcode_translator/executor/variables/container.py:
```diff
@@ -1123,29 +1125,39 @@ class SizeVariable(ListVariable):
-        self.graph.add_global_guarded_variable(key)
-        key_val = key.get_py_value()
+    @VariableFactory.register_from_value()
+    def from_value(value: Any, graph: FunctionGraph, tracker: Tracker):
```
`from_value` 永远放在最下面

LittleHeroZZZX replied:
Done

SigureMo commented on python/paddle/jit/sot/opcode_translator/executor/variables/container.py:
```diff
@@ -1123,29 +1125,39 @@ class SizeVariable(ListVariable):
-    def getitem(self, key):
-        self.graph.add_global_guarded_variable(key)
-        key_val = key.get_py_value()
+    @VariableFactory.register_from_value()
```
这里 `SizeVariable` 会在 `ListVariable` 后被找到吧？不加参数 `successor="ListVariable"` 没问题么？

LittleHeroZZZX replied:
Done

SigureMo commented on python/paddle/jit/sot/opcode_translator/executor/variables/container.py:
```diff
@@ -1123,29 +1125,39 @@ class SizeVariable(ListVariable):
-        key_val = key.get_py_value()
+    @VariableFactory.register_from_value()
+    def from_value(value: Any, graph: FunctionGraph, tracker: Tracker):
+        if type(value) is paddle.Size:
```
为什么不用 `isinstance(value, paddle.Size)`？

SigureMo replied:
这里还是用 `type` 吧，不能确定用户是否会自定义什么奇怪的子类

LittleHeroZZZX committed:
feat(sot): simplify dispatcher logic and add SizeVariable support

- Remove complex pattern scoring and candidate selection in Dispatcher._find_handler
- Add dispatcher registrations for paddle.Size conversions and operations
- Implement equality and inequality handlers for SizeVariable with tuple/list

LittleHeroZZZX committed:
feat(sot): remove unnecessary newline in Dispatcher.get_handler

The newline was removed for cleaner code formatting without affecting functionality. This improves readability in the pattern matching logic.

zrr1999 commented on python/paddle/jit/sot/opcode_translator/executor/variables/container.py:
这块也可以加一个 `assert self.__class__ in [ListVariable, SizeVariable]`

SigureMo replied:
`assert self.__class__ in {ListVariable, SizeVariable}`，set 就好

LittleHeroZZZX replied:
Done

zrr1999 commented on python/paddle/jit/sot/opcode_translator/executor/variables/container.py:
copy 是不是也应该用 `self.__class__` 来创建？

LittleHeroZZZX replied:
Done

zrr1999 commented on python/paddle/jit/sot/opcode_translator/executor/variables/container.py:
```diff
@@ -245,7 +247,7 @@ def getitem(self, key):
             return res
         elif isinstance(key, slice):
             items = self.proxy.get_all()
-            return VariableFactory.from_value(
+            return self.__class__(
                 items[key],
```
这块有个小问题，子类的__init__方法不一定跟父类有一样的方法签名，这里最安全的写法应该是用match或者ifelse列出所有可能的类型处理，其他未知情况报错。
不过现在的情况应该是没有影响的，因为已知所有子类都没有自定义init，所以可以先加一个 assert self.__class__ in [ListVariable, SizeVariable] 把现有所有情况包进去，能防止以后可能出现的问题。

SigureMo replied:
3.9 没退呢，哪来的 `match`

SigureMo commented on python/paddle/jit/sot/opcode_translator/executor/variable_dispatch.py:
```diff
@@ -263,6 +264,72 @@ def inner(*args, **kwargs):
+
+
+@Dispatcher.register_decorator(operator.eq)
+def dispatch_size_tuple_eq(
```
函数名应该改一下吧

LittleHeroZZZX replied:
Done，改成了 sequence

SigureMo commented on test/sot/test_25_size.py:
```diff
@@ -0,0 +1,204 @@
+from paddle.jit.sot.psdb import check_no_breakgraph, check_no_fallback
+
+
+@check_no_fallback
```
@check_no_fallback 可以删掉，`STRICT_MODE` 可以保证

gouzil replied:
文件名改成 `test_26_xxx` 吧

LittleHeroZZZX replied:
Done

LittleHeroZZZX committed:
feat(sot): enhance SizeVariable interoperability with sequence types

- Add support for concatenating List/Tuple/SizeVariable with SizeVariable using `operator.add`
- Rename equality/inequality dispatch functions for clarity (size_tuple -> size_sequence)
- Simplify SizeVariable constructor to directly accept List/TupleVariable arguments
- Ensure ListVariable methods (slice, repeat, copy) work correctly for SizeVariable subclass
- Remove redundant from_value factory method for SizeVariable

LittleHeroZZZX commented on test/sot/test_25_size.py:
```diff
@@ -0,0 +1,204 @@
+
+@check_no_breakgraph
+@check_no_fallback
+def size_add_list(x: int):
+    s = paddle.Size([x, 2])
+    l = [3, 4]
+    res = s + l
+    return res
+
+
+@check_no_fallback
+@check_no_breakgraph
+def size_add_tuple(x: int):
+    s = paddle.Size([x, 2])
+    t = (3, 4)
+    res = s + t
+    return res
+
+
+@check_no_fallback
+@check_no_breakgraph
+def size_add_size(x: int):
+    s1 = paddle.Size([x, 2])
+    s2 = paddle.Size([3, 4])
+    res = s1 + s2
+    return res
+
```
Done

LittleHeroZZZX commented on python/paddle/jit/sot/opcode_translator/executor/variable_dispatch.py:
```diff
@@ -263,6 +264,72 @@ def inner(*args, **kwargs):
+        tracker=DummyTracker([var]),
+    ),
+)
+
+Dispatcher.register(
+    operator.add,
+    ("SizeVariable", "ListVariable | TupleVariable | SizeVariable"),
+    lambda var, other: var.concat(other),
+)
+
+
```
Done, should convert var to size first, then call var.concat(other)

LittleHeroZZZX committed:
feat(sot): fix SizeVariable reconstruction and extend test coverage

- Change `push_null` from True to False in `SizeVariable._reconstruct` to correct code generation for paddle.Size
- Add test case for `size_numel` with zero dimension to improve coverage
- Update symbolic shape test to return both shape and numel for verification

SigureMo commented on test/sot/test_sot_paddle_size.py:
这个不是语法项，改成 `test_sot_paddle_size.py` 吧

LittleHeroZZZX replied:
Done

LittleHeroZZZX committed:
feat: rename test file for clarity and consistency

LittleHeroZZZX committed:
feat(sot): fix sequence equality comparison for SizeVariable

Previously, `dispatch_sequence_size_eq` incorrectly called `dispatch_size_sequence_eq` directly, which could lead to incorrect behavior when comparing a tuple/list to a SizeVariable. This change replaces the direct call with `Dispatcher.call(operator.eq, rhs, lhs)` to ensure proper dispatching and equality logic.

Updated test cases in `test_sot_paddle_size.py` to cover both directions of equality and inequality comparisons between paddle.Size and tuples/lists, verifying the fix works correctly for `==` and `!=` operators.
